pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -461,11 +468,57 @@ func (c *kvEventToRowConsumer) encodeAndEmit(
 		return err
 	}
 	if log.V(3) {
-		log.Infof(ctx, `r %s: %s -> %s`, updatedRow.TableName, keyCopy, valueCopy)
+		log.Infof(ctx, `r %s(%+v): %s -> %s`, updatedRow.TableName, keyCopy, headers, valueCopy)
 	}
 	return nil
 }
 
+var jsonHeaderWrongTypeLogLim = log.Every(1 * time.Minute)
+
+func (c *kvEventToRowConsumer) makeRowHeaders(
+	ctx context.Context, updatedRow cdcevent.Row,
+) (map[string][]byte, error) {
+	if c.encodingOpts.HeadersJSONColName == "" {
+		return nil, nil
+	}
+
+	headersJSON, err := readOneJSONValue(updatedRow, c.encodingOpts.HeadersJSONColName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read headers column %s: %s", c.encodingOpts.HeadersJSONColName, updatedRow.DebugString())
+	}
+	// Allow & ignore both kinds of null.
+	if headersJSON == nil || headersJSON.Type() == json.NullJSONType {
+		return nil, nil
+	}
+
+	headers := make(map[string][]byte, headersJSON.Len())
+	objIt, err := headersJSON.ObjectIter()
+	if err != nil {
+		return nil, err
+	}
+	if objIt == nil {
+		if jsonHeaderWrongTypeLogLim.ShouldLog() || log.V(2) {
+			log.Warningf(ctx, "headers column %s must be a JSON object, got %s - %s", c.encodingOpts.HeadersJSONColName, headersJSON.Type(), headersJSON)
+		}
+		return nil, nil
+	}
+	// Format the object's kv pairs. The intended use case is to have string
+	// values, but permit other types, formatted as JSON. I'd rather have
+	// confusing headers than the feed choke on specific values.
```

Comment:
Has this behavior been passed by @rohan-joshi ? Will users potentially be confused because the format being wrong might imply that the header value itself is wrong?

Reply:
I would like his feedback on it -- my feeling is that it's better to give the user the benefit of the doubt rather than irrevocably fail the changefeed